# Notebook 1: Exploratory Data Analysis (EDA)

In this notebook I explore the music dataset to understand the data before building a model.
I look at distributions, correlations, and patterns across all the features.

In [ ]:
# Importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

# Load the training dataset (one level up from notebooks/ folder)
df = pd.read_csv('../train.csv')

# Quick look at shape and first few rows
print("Shape of dataset:", df.shape)
df.head()

## 1. Basic Data Overview

First, I check what columns we have, their types, and if there are any missing values.

In [11]:
# Check data types and missing values
print("Data types:")
print(df.dtypes)
print()
print("Missing values per column:")
print(df.isnull().sum())

Data types:
Unnamed: 0                    int64
artist_name                  object
track_name                   object
release_date                  int64
genre                        object
lyrics                       object
len                           int64
dating                      float64
violence                    float64
world/life                  float64
night/time                  float64
shake the audience          float64
family/gospel               float64
romantic                    float64
communication               float64
obscene                     float64
music                       float64
movement/places             float64
light/visual perceptions    float64
family/spiritual            float64
sadness                     float64
feelings                    float64
topic                        object
age                         float64
dtype: object

Missing values per column:
Unnamed: 0                  0
artist_name                 0
track_name            

In [12]:
# Summary statistics for numeric columns
df.describe()

,Unnamed: 0,release_date,len,dating,violence,world/life,night/time,shake the audience,family/gospel,romantic,communication,obscene,music,movement/places,light/visual perceptions,family/spiritual,sadness,feelings,age
count,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000,28362.000000
mean,42948.166878,1990.239652,73.030534,0.021110,0.118371,0.120984,0.057356,0.017418,0.017045,0.048676,0.076651,0.097185,0.060067,0.047417,0.049008,0.024155,0.129402,0.030995,0.425148
std,24747.811462,18.486997,41.831605,0.052366,0.178658,0.172216,0.111892,0.040658,0.041968,0.106071,0.109497,0.181314,0.123346,0.091559,0.089553,0.051032,0.181149,0.071656,0.264100
min,0.000000,1950.000000,1.000000,0.000291,0.000284,0.000291,0.000289,0.000284,0.000289,0.000284,0.000291,0.000289,0.000289,0.000284,0.000284,0.000284,0.000284,0.000289,0.014286
25%,20393.500000,1975.000000,42.000000,0.000923,0.001120,0.001170,0.001032,0.000993,0.000923,0.000975,0.001144,0.001053,0.000975,0.000993,0.000993,0.000957,0.001144,0.000993,0.185714
50%,45407.000000,1991.000000,63.000000,0.001462,0.002506,0.006579,0.001949,0.001595,0.001504,0.001754,0.002632,0.001815,0.001815,0.001645,0.001815,0.001645,0.005263,0.001754,0.414286
75%,64089.500000,2007.000000,93.000000,0.004049,0.192538,0.197854,0.065778,0.009989,0.004785,0.042304,0.132111,0.088799,0.055109,0.054373,0.064302,0.025515,0.235115,0.032617,0.642857
max,82451.000000,2019.000000,199.000000,0.647706,0.981781,0.962105,0.973684,0.497463,0.545303,0.940789,0.645829,0.992298,0.956938,0.638021,0.667782,0.618073,0.981424,0.958810,1.000000


## 2. Genre Distribution

How many songs fall into each genre? This tells me if the dataset is balanced.

In [ ]:
# Count songs per genre and plot
plt.figure(figsize=(10, 5))
genre_counts = df['genre'].value_counts()
sns.barplot(x=genre_counts.index, y=genre_counts.values, hue=genre_counts.index, palette='viridis', legend=False)
plt.title('Number of Songs per Genre')
plt.xlabel('Genre')
plt.ylabel('Count')
plt.xticks(rotation=45)
plt.tight_layout()
os.makedirs('../docs', exist_ok=True)
plt.savefig('../docs/genre_distribution.png')
plt.show()
print(genre_counts)

**What this chart shows:** This is a bar chart showing how many songs belong to each genre. I chose this because the first thing I wanted to know before doing any analysis is what kind of data I'm working with, specifically whether some genres have way more songs than others.

**What I found:** Pop has by far the most songs in the dataset, followed by country and rock. Genres like jazz and reggae have significantly fewer songs. This tells me the dataset is imbalanced, meaning the clustering model might end up being more influenced by pop songs just because there are more of them.

## 3. Distribution of Lyrical Feature Scores

Each song has scores from 0 to 1 for features like dating, violence, romantic, etc.
Let me look at the distribution of each one.

In [ ]:
# Define the lyrical feature columns
lyric_features = [
    'dating', 'violence', 'world/life', 'night/time',
    'shake the audience', 'family/gospel', 'romantic',
    'communication', 'obscene', 'music', 'movement/places',
    'light/visual perceptions', 'family/spiritual', 'sadness', 'feelings'
]

# Plot histogram for each feature
fig, axes = plt.subplots(5, 3, figsize=(15, 18))
axes = axes.flatten()

for i, col in enumerate(lyric_features):
    axes[i].hist(df[col], bins=30, color='steelblue', edgecolor='white')
    axes[i].set_title(col)
    axes[i].set_xlabel('Score')
    axes[i].set_ylabel('Frequency')

plt.suptitle('Distribution of Lyrical Feature Scores', fontsize=16, y=1.01)
plt.tight_layout()
plt.savefig('../docs/feature_distributions.png')
plt.show()

**What this chart shows:** These are histograms, one for each lyrical feature score. A histogram shows how the values are spread out. I wanted to see whether songs are spread evenly across the 0-1 range or if most songs cluster at one end.

**What I found:** Almost every feature is heavily skewed to the left meaning the vast majority of songs score very close to 0. For example, most songs score near 0 on violence or gospel. Only a small number of songs score high on any given feature. This makes sense because most songs are not specifically about one theme. This skewness is something to keep in mind when clustering.

## 4. Correlation Heatmap

Now I want to see which features are strongly correlated.
If two features are very correlated, we might not need both in our model.

In [ ]:
# Compute correlation matrix for lyrical features + age + len
numeric_cols = lyric_features + ['len', 'age']
corr_matrix = df[numeric_cols].corr()

# Plot heatmap
plt.figure(figsize=(14, 10))
sns.heatmap(
    corr_matrix,
    annot=True,
    fmt='.2f',
    cmap='coolwarm',
    center=0,
    square=True,
    linewidths=0.5
)
plt.title('Correlation Heatmap of Numeric Features')
plt.tight_layout()
plt.savefig('../docs/correlation_heatmap.png')
plt.show()

**What this chart shows:** This is a correlation heatmap. It shows how related each pair of features is to each other. A value close to 1 (dark red) means two features tend to go up and down together. A value close to -1 (dark blue) means they move in opposite directions. A value close to 0 means they are not related.

**What I found:** Most features are not strongly correlated with each other. No pair of columns exceeded a 0.85 correlation, so I did not need to drop any columns based on this. There is a mild positive relationship between romantic and dating, which makes sense, but it was not strong enough to remove either column.

## 5. Average Lyrical Scores by Genre

This shows how each genre scores across different lyrical themes.
It helps understand what makes each genre unique.

In [ ]:
# Group by genre and compute mean of each lyric feature
genre_means = df.groupby('genre')[lyric_features].mean()

# Heatmap of genre vs feature means
plt.figure(figsize=(16, 6))
sns.heatmap(
    genre_means,
    annot=True,
    fmt='.2f',
    cmap='YlOrRd',
    linewidths=0.5
)
plt.title('Average Lyrical Feature Scores by Genre')
plt.xlabel('Feature')
plt.ylabel('Genre')
plt.tight_layout()
plt.savefig('../docs/genre_feature_heatmap.png')
plt.show()

**What this chart shows:** This heatmap shows the average score of each lyrical feature broken down by genre. I chose this because I wanted to see if different genres have different lyrical personalities, and this chart makes that really easy to spot.

**What I found:** Yes, genres are clearly different from each other lyrically. For example, reggae scores high on world/life, rock scores higher on violence, and pop leans more toward romantic and dating themes. This was one of the most useful charts in the EDA because it directly shaped my hypothesis, if genres already have distinct lyrical profiles, then the clustering algorithm should be able to find similar groupings on its own.

## 6. Bivariate: Romantic vs. Dating Score

Intuitively, songs about romance and dating might be correlated.

In [ ]:
# Scatter plot: romantic vs dating colored by genre
plt.figure(figsize=(9, 6))
genres_list = df['genre'].unique()
colors = sns.color_palette('tab10', len(genres_list))

for genre, color in zip(genres_list, colors):
    subset = df[df['genre'] == genre]
    plt.scatter(subset['romantic'], subset['dating'],
                alpha=0.4, label=genre, color=color, s=10)

plt.xlabel('Romantic Score')
plt.ylabel('Dating Score')
plt.title('Romantic vs. Dating Score by Genre')
plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', fontsize=8)
plt.tight_layout()
plt.savefig('../docs/romantic_vs_dating.png')
plt.show()

**What this chart shows:** This is a scatter plot comparing each song's romantic score (x-axis) against its dating score (y-axis), with each genre shown in a different color. I chose this specific pair because I suspected these two features might be redundant songs about romance often involve dating too.

**What I found:** There is a slight positive trend, meaning songs that score high on romantic do tend to score a bit higher on dating as well. However, the relationship is not strong enough to say they are measuring the same thing. I decided to keep both columns.

## 7. Song Length (Word Count) by Genre

In [ ]:
# Boxplot of word count by genre
plt.figure(figsize=(12, 5))
df.boxplot(column='len', by='genre', figsize=(12, 5))
plt.suptitle('')
plt.title('Song Word Count by Genre')
plt.xlabel('Genre')
plt.ylabel('Number of Words')
plt.xticks(rotation=45)
plt.tight_layout()
plt.savefig('../docs/wordcount_by_genre.png')
plt.show()

**What this chart shows:** This is a boxplot showing the distribution of word counts (how many words are in the lyrics) for each genre. The box shows where the middle 50% of songs fall, and the dots outside the whiskers are outliers.

**What I found:** Hip hop and reggae songs tend to have more words than other genres, which makes a lot of sense those genres rely heavily on dense, wordy lyrics. Jazz and pop songs tend to be shorter. There are also some extreme outliers in most genres, which is why I removed the top 1% of songs by word count during preprocessing to avoid those skewing the model.

## EDA Summary

- The dataset has **28,362 songs** across multiple genres with **no missing values**.
- Most lyrical feature scores are right-skewed — most songs score low on things like violence or gospel.
- No pair of columns exceeded a **0.85 correlation threshold**, so no columns needed to be dropped based on correlation alone.
- There is a mild positive relationship between `romantic` and `dating`, but not strong enough to remove either.
- Different genres have distinct lyrical profiles — for example, gospel scores high on `family/spiritual`, while rock scores higher on `violence`.
- Song length varies quite a bit across genres, with hip hop and reggae having the longest lyrics.

These observations will guide preprocessing decisions in the next notebook.

## Hypothesis

Looking at the data, I think songs will group together based on their lyrical themes more than anything else. From the genre heatmap, it's pretty clear that different genres have very different lyrical profiles, for example, rock songs tend to score higher on violence while gospel songs score high on family/spiritual. So my hypothesis is that when we run the clustering algorithm, the clusters will end up reflecting these kinds of theme differences.

I also noticed that romantic and dating scores seem to go together, and sadness shows up a lot across genres. Because of that, I think we'll probably see some clusters centered around emotional themes like sadness or romance, and others around more aggressive or spiritual content.

Basically, I'm expecting the algorithm to find groups that make intuitive sense ,like a "love song" cluster, a "rock/aggressive" cluster, and a "gospel/spiritual" cluster ,even though we're not giving it any labels to work with.